File: backend/app/models/observability_event.py
"""ORM model for the ``observability_events`` table.

Stores structured observability events uploaded from the Android app via
``POST /api/v1/observability/events`` (every 15 minutes via WorkManager).

These events are the primary data source for Phase 10 AI Error Analysis.
The analysis pipeline reads recent ERROR/CRITICAL events for a given
session or time window and passes them to the LLM alongside retrieved
runbooks and historical incident reports.

Fields mirror the Android ``ObservabilityEvent`` data class defined in
``core-common/src/main/kotlin/.../observability/ObservabilityEvent.kt``.

Phase 10 — AI Error Analysis
"""

from __future__ import annotations

import uuid
from datetime import datetime

from sqlalchemy import BigInteger, DateTime, Index, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import Base, uuid_pk


class ObservabilityEvent(Base):
    """Single observability event uploaded from the Android app."""

    __tablename__ = "observability_events"

    # ── Primary key ───────────────────────────────────────────────────────────
    id: Mapped[uuid.UUID] = uuid_pk()

    # ── Event identity ────────────────────────────────────────────────────────
    # epoch millis (UTC) when the event was captured on-device
    timestamp_ms: Mapped[int] = mapped_column(
        BigInteger,
        nullable=False,
        comment="Epoch milliseconds (UTC) when the event was captured on-device",
    )

    # DEBUG | INFO | WARN | ERROR | CRITICAL
    level: Mapped[str] = mapped_column(
        String(16),
        nullable=False,
        index=True,
        comment="Event severity: DEBUG | INFO | WARN | ERROR | CRITICAL",
    )

    # e.g. "network_error", "crash_handled", "api_latency"
    event_type: Mapped[str] = mapped_column(
        String(128),
        nullable=False,
        index=True,
        comment="Machine-readable event category (use EventType constants)",
    )

    # PII-filtered human-readable description
    message: Mapped[str] = mapped_column(
        Text,
        nullable=False,
        comment="Human-readable, PII-stripped event description",
    )

    # ── Correlation identifiers ───────────────────────────────────────────────
    session_id: Mapped[str] = mapped_column(
        String(128),
        nullable=False,
        index=True,
        comment="Groups all events within one app session (process lifetime UUID)",
    )

    request_id: Mapped[str | None] = mapped_column(
        String(128),
        nullable=True,
        index=True,
        comment="Unique per HTTP call — correlates Android log with backend log",
    )

    trace_id: Mapped[str | None] = mapped_column(
        String(128),
        nullable=True,
        comment="Groups related events across a single user action flow",
    )

    # ── Context ───────────────────────────────────────────────────────────────
    # Compose navigation route active when the event fired
    screen: Mapped[str | None] = mapped_column(
        String(256),
        nullable=True,
        comment="Active Compose screen/route at time of event",
    )

    # JSON-encoded metadata dict (all values are strings, no PII)
    metadata_json: Mapped[str] = mapped_column(
        Text,
        nullable=False,
        default="{}",
        comment="JSON-encoded Map<String,String> of event metadata",
    )

    # ── Ingest timestamp ──────────────────────────────────────────────────────
    # When the backend received this event (not when it was captured on-device)
    received_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
        index=True,
        comment="UTC timestamp when this event was received by the backend",
    )

    # ── Composite indexes for common query patterns ───────────────────────────
    __table_args__ = (
        # Most common query: recent ERROR/CRITICAL events for analysis
        Index("ix_obs_evt_level_received", "level", "received_at"),
        # Correlate a full session in one query
        Index("ix_obs_evt_session_received", "session_id", "received_at"),
        # Look up all events for a specific request
        Index("ix_obs_evt_request_id", "request_id"),
    )

    def __repr__(self) -> str:
        return (
            f"<ObservabilityEvent id={self.id!s} level={self.level!r} "
            f"event_type={self.event_type!r} received_at={self.received_at!s}>"
        )

File: core-common/src/main/kotlin/com/aiassistant/core/common/observability/ObservabilityEvent.kt
/**
 * ObservabilityEvent.kt — core-common module
 *
 * Purpose: Standard event model for all observability data captured by the Android app.
 *          Every meaningful runtime event — network call, crash, screen view, lifecycle
 *          change — is represented as an [ObservabilityEvent] before being forwarded to
 *          the backend AI analysis pipeline.
 *
 * Architecture: core-common — pure Kotlin, zero Android/framework dependencies.
 *               This type is the single shared currency between the capture layer
 *               (interceptors, crash handlers) and the storage/upload layer (WorkManager).
 *
 * Design decisions:
 * - [metadata] is a Map<String, String> (not Any) so the entire event is safely
 *   JSON-serializable without reflection and never leaks complex objects.
 * - [traceId] groups all events belonging to one logical user action (e.g. "tap Send →
 *   POST /chat → streaming response"). When null, the event is stand-alone.
 * - [requestId] is unique per HTTP call — used to correlate the Android log entry with
 *   the backend log entry for the same request.
 * - [sessionId] groups all events within a single app launch-to-close session.
 * - PII is never stored raw in this model. The [PiiFilter] must be applied before
 *   constructing an [ObservabilityEvent].
 *
 * Phase 2 — Android Observability
 */

package com.aiassistant.core.common.observability

import kotlinx.serialization.Serializable

// ─── Event level ─────────────────────────────────────────────────────────────

/**
 * Severity / importance of an [ObservabilityEvent].
 *
 * Mirrors standard log-level semantics so backend consumers can filter by severity
 * without parsing the event body.
 */
enum class EventLevel {
    /** Verbose diagnostic output — not sent to the backend in production builds. */
    DEBUG,

    /** Routine operational events (screen opened, session started). */
    INFO,

    /** Degraded but recoverable state (slow API, retried request). */
    WARN,

    /** A user-visible failure occurred (API error, crash handled). */
    ERROR,

    /** A crash or data-loss event that requires immediate attention. */
    CRITICAL
}

// ─── Event type constants ─────────────────────────────────────────────────────

/**
 * Well-known values for [ObservabilityEvent.eventType].
 *
 * Using string constants (rather than an enum) keeps the model open for extension:
 * any module can define its own event types without modifying this file.
 */
object EventType {
    // Network
    const val NETWORK_ERROR = "network_error"
    const val NETWORK_TIMEOUT = "network_timeout"
    const val API_LATENCY = "api_latency"
    const val HTTP_ERROR = "http_error" // 4xx / 5xx responses
    const val HTTP_SUCCESS = "http_success"

    // Crashes & exceptions
    const val CRASH_UNHANDLED = "crash_unhandled"
    const val CRASH_HANDLED = "crash_handled"

    // Lifecycle
    const val APP_FOREGROUND = "app_foreground"
    const val APP_BACKGROUND = "app_background"
    const val SCREEN_VIEW = "screen_view"

    // Session
    const val SESSION_START = "session_start"
    const val SESSION_END = "session_end"

    // User-visible errors
    const val USER_ERROR = "user_error"
}

// ─── Core event model ─────────────────────────────────────────────────────────

/**
 * Immutable record describing a single observable event within the Android app.
 *
 * ### Required fields
 * - [timestamp] — epoch millis (UTC) when the event was captured.
 * - [level] — severity used for routing and alerting.
 * - [eventType] — machine-readable category; use constants from [EventType].
 * - [message] — human-readable description, PII-stripped.
 * - [sessionId] — groups events within one app session (set by [SessionManager]).
 *
 * ### Optional correlation fields
 * - [screen] — active Compose destination name at capture time.
 * - [requestId] — unique per HTTP call; links Android log ↔ backend log.
 * - [traceId] — groups related events across a single user flow.
 *
 * ### Extensible context
 * - [metadata] — arbitrary key-value pairs; all values are strings to guarantee
 *   safe serialization. No PII may appear here.
 *
 * Example:
 * ```kotlin
 * ObservabilityEvent(
 *     timestamp   = System.currentTimeMillis(),
 *     level       = EventLevel.ERROR,
 *     eventType   = EventType.HTTP_ERROR,
 *     message     = "POST /chat returned HTTP 500",
 *     screen      = "ChatScreen",
 *     requestId   = UUID.randomUUID().toString(),
 *     traceId     = currentTraceId,
 *     sessionId   = sessionManager.sessionId,
 *     metadata    = mapOf(
 *         "http_status" to "500",
 *         "endpoint"    to "/chat",
 *         "latency_ms"  to "320",
 *     ),
 * )
 * ```
 */
@Serializable
data class ObservabilityEvent(
    /** Epoch milliseconds (UTC) when the event was captured on-device. */
    val timestamp: Long,

    /** Severity level — used by the backend to filter and route events. */
    val level: EventLevel,

    /**
     * Machine-readable event category.
     * Use one of the [EventType] constants or a module-specific string.
     */
    val eventType: String,

    /**
     * Human-readable description of what happened.
     * Must be PII-free — apply [PiiFilter] before populating this field.
     */
    val message: String,

    /**
     * Compose navigation route / screen name that was active when the event fired.
     * Null for events not tied to a specific screen (e.g. background syncs).
     */
    val screen: String? = null,

    /**
     * Unique identifier for a single HTTP request.
     * Correlates this Android log entry with the corresponding backend log line.
     * Null for non-network events.
     */
    val requestId: String? = null,

    /**
     * Trace ID that groups all events belonging to one logical user action.
     * Example: one [traceId] spans "tap Send → POST /chat → stream chunks → render".
     * Null when no trace context has been established.
     */
    val traceId: String? = null,

    /**
     * Session ID assigned at app launch by [SessionManager].
     * Groups all events within a single foreground-to-background lifecycle.
     */
    val sessionId: String,

    /**
     * Arbitrary key-value metadata for this event.
     * All values must be plain strings — no nested objects, no PII.
     *
     * Common keys:
     * - `"http_status"`   — HTTP response code as a string, e.g. `"500"`
     * - `"endpoint"`      — API path, e.g. `"/chat"`
     * - `"latency_ms"`    — Request duration in milliseconds, e.g. `"320"`
     * - `"error_type"`    — DomainError subtype name, e.g. `"NetworkError"`
     * - `"retry_count"`   — Number of retries attempted, e.g. `"2"`
     */
    val metadata: Map<String, String> = emptyMap()
)
